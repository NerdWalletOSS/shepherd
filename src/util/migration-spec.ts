import Joi from 'joi';
import fs from 'fs';
import * as yaml from 'js-yaml';
import { cloneDeep, mapValues } from 'lodash';
import path from 'path';

export interface IMigrationHooks {
  should_migrate?: string[];
  should_create_issue?: string[];
  post_checkout?: string[];
  apply?: string[];
  pr_message?: string[];
  [name: string]: string[] | undefined;
}

export interface IMigrationIssues {
  title?: string | number;
  description?: string;
  labels?: string[];
  state?: 'open' | 'closed';
  state_reason?: 'completed' | 'not_planned' | 'reopened' | null;
}

export type MigrationPhase = [keyof IMigrationHooks];

export interface IMigrationSpec {
  id: string;
  title: string;
  adapter: {
    type: string;
    [key: string]: any;
  };
  hooks: IMigrationHooks;
  issues?: IMigrationIssues;
}

export function loadSpec(directory: string): IMigrationSpec {
  const docPath = path.join(directory, 'shepherd.yml');
  const spec = yaml.load(fs.readFileSync(docPath, 'utf8'));
  const normalizedSpec = normalizeSpec(spec);
  const validationResult = validateSpec(normalizedSpec);
  if (validationResult.error) {
    throw new Error(`Error loading migration spec: ${validationResult.error.message}`);
  }
  return normalizedSpec;
}

export function normalizeSpec(originalSpec: any): IMigrationSpec {
  const spec = cloneDeep(originalSpec);
  if (spec.hooks) {
    spec.hooks = mapValues(spec.hooks, (steps: any, phase: string) => {
      if (typeof steps === 'string') {
        return [steps];
      } else if (Array.isArray(steps)) {
        return steps;
      } else {
        throw new Error(`Error reading shepherd.yml: ${phase} must be a string or array`);
      }
    });
  } else {
    spec.hooks = {};
  }
  return spec;
}

export function validateSpec(spec: any) {
  const hookSchema = Joi.array().items(Joi.string());
  const schema = Joi.object({
    id: Joi.string().required(),
    title: Joi.string().required(),
    adapter: Joi.object({
      type: Joi.string().valid('github').required(),
    })
      .unknown(true)
      .required(),
    hooks: Joi.object({
      should_migrate: hookSchema,
      should_create_issue: hookSchema,
      post_checkout: hookSchema,
      apply: hookSchema,
      pr_message: hookSchema,
    }),
    issues: Joi.object({
      title: Joi.any().required(),
      description: Joi.string().required(),
      state: Joi.string().optional(),
      state_reason: Joi.string().optional(),
      labels: hookSchema.optional(),
    }).optional(),
  });

  return schema.validate(spec);
}
