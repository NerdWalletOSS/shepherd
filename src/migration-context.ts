import IRepoAdapter, { IRepo } from './adapters/base.js';
import { ILogger } from './logger/index.js';
import { IMigrationSpec } from './util/migration-spec.js';

export interface IShepherdInfo {
  workingDirectory: string;
}

export interface IMigrationInfo {
  spec: IMigrationSpec;
  migrationDirectory: string;
  workingDirectory: string;
  repos: IRepo[] | null;
  upstreamOwner: string;
  selectedRepos?: IRepo[];
}

export interface IMigrationContext {
  shepherd: IShepherdInfo;
  migration: IMigrationInfo;
  adapter: IRepoAdapter;
  logger: ILogger;
}

export enum state {
  open,
  closed,
}

export enum state_reason {
  completed,
  not_planned,
  reopened,
  null,
}
