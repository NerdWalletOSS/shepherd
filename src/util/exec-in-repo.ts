import { spawn } from 'child-process-promise';
import { IRepo } from '../adapters/base.js';
import { IMigrationContext } from '../migration-context.js';

export default async (context: IMigrationContext, repo: IRepo, command: string): Promise<any> => {
  const repoDir = context.adapter.getRepoDir(repo);
  const dataDir = context.adapter.getDataDir(repo);
  const baseBranch = context.adapter.getBaseBranch(repo);
  const migrationDir = context.migration.migrationDirectory;
  const adapterEnvironmentVars = await context.adapter.getEnvironmentVariables(repo);

  const execOptions = {
    cwd: repoDir,
    env: {
      ...process.env,
      SHEPHERD_REPO_DIR: repoDir,
      SHEPHERD_DATA_DIR: dataDir,
      SHEPHERD_MIGRATION_DIR: migrationDir,
      SHEPHERD_BASE_BRANCH: baseBranch,
      ...adapterEnvironmentVars,
    },
    shell: true,
    capture: ['stdout', 'stderr'],
  };
  return spawn(command, [], execOptions);
};
