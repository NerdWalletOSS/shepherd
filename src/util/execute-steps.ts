import chalk from 'chalk';
import { IRepo } from '../adapters/base.js';
import { IMigrationContext } from '../migration-context.js';
import execInRepo from '../util/exec-in-repo.js';

interface IStepResult {
  step: string;
  succeeded: boolean;
  stdout?: string;
  stderr?: string;
}

export interface IStepsResults {
  succeeded: boolean;
  stepResults: IStepResult[];
}

const executeStep = async (
  context: IMigrationContext,
  repo: IRepo,
  step: string,
  showOutput: boolean,
  repoLogs: string[]
): Promise<IStepResult> => {
  repoLogs.push(`\$ ${step}`);

  try {
    const { stdout, stderr } = await execInRepo(context, repo, step);

    if (showOutput) {
      repoLogs.push(stdout);
    }

    repoLogs.push(stderr);
    repoLogs.push(chalk.green(`Step "${step}" exited with 0`));

    return {
      step,
      succeeded: true,
      stdout: stdout,
      stderr: stderr,
    };
  } catch (e: any) {
    if (e.code !== undefined) {
      repoLogs.push(`Step "${step}" exited with ${e.code}`);
    } else {
      repoLogs.push(e.toString());
    }

    return {
      step,
      succeeded: false,
      stdout: e.stdout,
      stderr: e.stderr,
    };
  }
};

/**
 * Executes a series of migration steps for a given phase.
 *
 * @param context - The migration context containing necessary information and configurations.
 * @param repo - The repository interface to interact with the repository.
 * @param phase - The phase of the migration to execute steps for.
 * @param showOutput - Optional flag to determine if output should be shown. Defaults to true.
 * @param repoLogs - Array to collect logs from the repository.
 * @returns A promise that resolves to the results of the executed steps.
 */
export default async (
  context: IMigrationContext,
  repo: IRepo,
  phase: string,
  showOutput = true,
  repoLogs: string[] = []
): Promise<IStepsResults> => {
  const {
    migration: {
      spec: { hooks },
    },
  } = context;

  const results: IStepsResults = {
    succeeded: false,
    stepResults: [],
  };

  const steps = hooks[phase] || [];

  for (const step of steps) {
    const stepResult = await executeStep(context, repo, step, showOutput, repoLogs);
    results.stepResults.push(stepResult);

    if (!stepResult.succeeded) {
      return results;
    }
  }

  results.succeeded = true;
  return results;
};
