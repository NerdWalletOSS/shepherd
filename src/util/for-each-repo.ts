import chalk from 'chalk';
import fs from 'fs-extra';
import { IRepo } from '../adapters/base.js';
import { IMigrationContext } from '../migration-context.js';

type RepoHandler = (repo: IRepo, header: string) => Promise<void>;

interface IOptions {
  warnMissingDirectory?: boolean;
}

const getRepos = (migrationRepos: IRepo[], selectedRepos: IRepo[], adapter: any): IRepo[] => {
  if (selectedRepos && selectedRepos.length) {
    return selectedRepos.map((r) => {
      const existingRepo = migrationRepos.find((repo) => adapter.reposEqual(r, repo));
      return existingRepo || r;
    });
  }
  return migrationRepos || [];
};

const logRepoInfo = (repo: IRepo, index: number, total: number, adapter: any) => {
  const indexString = chalk.dim(`${index}/${total}`);
  return chalk.bold(`\n[${adapter.stringifyRepo(repo)}] ${indexString}`);
};

const checkRepoDirectory = async (repoDir: string, warnMissingDirectory: boolean) => {
  if (warnMissingDirectory && !(await fs.pathExists(repoDir))) {
    return `Directory ${repoDir} does not exist`;
  }
  return '';
};

/**
 * Processes each repository in the migration context using the provided handler.
 *
 * @param context - The migration context containing repositories, logger, and adapter.
 * @param param1 - Either a repository handler function or options for processing.
 * @param param2 - Optional repository handler function if `param1` is options.
 *
 * @returns A promise that resolves when all repositories have been processed.
 *
 * The function extracts repositories from the migration context and applies the provided handler
 * to each repository. It logs warnings for missing directories and errors encountered during processing.
 *
 * @example
 * ```typescript
 * await forEachRepo(context, handler);
 * await forEachRepo(context, options, handler);
 * ```
 */
export default async (
  context: IMigrationContext,
  param1: RepoHandler | IOptions,
  param2?: RepoHandler
) => {
  const {
    migration: { repos: migrationRepos, selectedRepos },
    logger,
    adapter,
  } = context;

  const { handler, options } =
    typeof param1 === 'function'
      ? { handler: param1, options: {} as IOptions }
      : { handler: param2 as RepoHandler, options: param1 };

  const { warnMissingDirectory = true } = options;
  const repos = getRepos(migrationRepos || [], selectedRepos || [], adapter);

  const _handler = async function (repo: IRepo, index: number, repos: IRepo[]) {
    const repoDir = adapter.getRepoDir(repo);
    const dirCheckMessage = await checkRepoDirectory(repoDir, warnMissingDirectory);
    if (dirCheckMessage) {
      logger.warn(dirCheckMessage);
    }

    try {
      const header = logRepoInfo(repo, index + 1, repos.length, adapter);
      await handler(repo, header);
    } catch (error: unknown) {
      if (error instanceof Error) {
        logger.error(`Error processing repo ${adapter.stringifyRepo(repo)}: ${error.message}`);
      } else {
        logger.error(`Error processing repo ${adapter.stringifyRepo(repo)}: ${String(error)}`);
      }
    }
  };

  const repoListWithHandlersApplied = repos.map(_handler);

  await Promise.all(repoListWithHandlersApplied);
};
