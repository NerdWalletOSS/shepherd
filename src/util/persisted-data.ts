import fs from 'fs-extra';
import { load } from 'js-yaml';
import _ from 'lodash';
const { differenceWith, unionWith } = _;
import path from 'path';

import IRepoAdapter, { IRepo, IssueTracker } from '../adapters/base.js';
import { IMigrationContext } from '../migration-context.js';

const jsonStringify = (data: any) => JSON.stringify(data, undefined, 2);

/**
 * This is to support people who started using Shepherd before we switched to
 * storing state in JSON. It'll read an existing YAML file, write it to a JSON
 * file, and delete the old YAML file.
 */
const migrateToJsonIfNeeded = async (migrationContext: IMigrationContext) => {
  const legacyFile = getLegacyRepoListFile(migrationContext);
  if (await fs.pathExists(legacyFile)) {
    const data = load(await fs.readFile(legacyFile, 'utf8'));
    await fs.outputFile(getRepoListFile(migrationContext), jsonStringify(data));
    await fs.remove(legacyFile);
  }
};

const getRepoListFile = (migrationContext: IMigrationContext) => {
  return path.join(migrationContext.migration.workingDirectory, 'repos.json');
};

const getIssueTrackerFile = (migrationContext: IMigrationContext) => {
  return path.join(migrationContext.migration.workingDirectory, 'issue_tracker.json');
};

export const removeRepoDirectories = async (adapter: IRepoAdapter, repo: IRepo) => {
  await fs.remove(adapter.getRepoDir(repo));
  await fs.remove(adapter.getDataDir(repo));
};

const getIssueListsFromTracker = async (migrationContext: IMigrationContext) => {
  const issueTrackerFile = getIssueTrackerFile(migrationContext);

  if (await fs.pathExists(issueTrackerFile)) {
    return JSON.parse(await fs.readFile(issueTrackerFile, 'utf8'));
  }
  return [];
};

const getLegacyRepoListFile = (migrationContext: IMigrationContext) => {
  return path.join(migrationContext.migration.workingDirectory, 'repos.yml');
};

const loadRepoList = async (migrationContext: IMigrationContext): Promise<IRepo[] | null> => {
  await migrateToJsonIfNeeded(migrationContext);
  const repoListFile = getRepoListFile(migrationContext);
  if (!(await fs.pathExists(repoListFile))) {
    return null;
  }
  return JSON.parse(await fs.readFile(repoListFile, 'utf8'));
};

const updateRepoList = async (
  migrationContext: IMigrationContext,
  checkedOutRepos: IRepo[],
  discardedRepos: IRepo[]
): Promise<IRepo[]> => {
  // We need to keep the list of repos in sync with what's actually on disk
  // To do this, we'll load the existing list, delete any repos that were not
  // kept during the checkout process (perhaps they failed a new should_migrate check),
  // and add any repos that were newly checked out, removing duplicates as appropriate
  const existingRepos = await loadRepoList(migrationContext);
  if (!existingRepos) {
    // No repos stored yet, we can update this list directly
    await fs.outputFile(getRepoListFile(migrationContext), JSON.stringify(checkedOutRepos));
    return checkedOutRepos;
  }

  const { reposEqual } = migrationContext.adapter;
  const repos = unionWith(
    differenceWith(existingRepos, discardedRepos, reposEqual),
    checkedOutRepos,
    reposEqual
  );

  await fs.outputFile(getRepoListFile(migrationContext), JSON.stringify(repos));
  return repos;
};

const updatePostedIssuesLists = async (
  migrationContext: IMigrationContext,
  issuePostedRepos: IssueTracker[]
): Promise<void> => {
  const issueTrackerFile = getIssueTrackerFile(migrationContext);
  //write all posted issues with issue number in the issue_tracker.json
  return await fs.outputFile(issueTrackerFile, JSON.stringify(issuePostedRepos));
};

const getIssueNumberForRepo = async (
  migrationContext: IMigrationContext,
  repo: string
): Promise<any> => {
  const issueTrackerFile = getIssueTrackerFile(migrationContext);
  if (!(await fs.pathExists(issueTrackerFile))) {
    return null;
  }
  const issueTracker: IssueTracker[] = JSON.parse(await fs.readFile(issueTrackerFile, 'utf8'));
  // read the issue number from the issue tracker for each repo
  return issueTracker
    .filter((issue) => issue.repo === repo)
    .map((specificRepo) => specificRepo.issueNumber);
};

export {
  updateRepoList,
  loadRepoList,
  updatePostedIssuesLists,
  getIssueTrackerFile,
  getIssueListsFromTracker,
  getIssueNumberForRepo,
};
