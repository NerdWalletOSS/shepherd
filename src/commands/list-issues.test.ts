import listIssues from './list-issues';
import { IMigrationContext } from '../migration-context';
import mockAdapter from '../adapters/adapter.mock';
import mockLogger from '../logger/logger.mock';
import { getIssueTrackerFile } from '../util/persisted-data';

jest.mock('fs-extra', () => {
    return {
        // Mock other methods as needed
        readFile: jest.fn().mockResolvedValue('{"name": "test"}'),
    };
});

jest.mock('../util/persisted-data');

describe('list-issue command', () => {
    let mockContext: IMigrationContext;

    beforeEach(() => {
        jest.clearAllMocks();

        mockContext = {
            shepherd: {
                workingDirectory: 'workingDirectory',
            },
            migration: {
                migrationDirectory: 'migrationDirectory',
                spec: {
                    id: 'id',
                    title: 'title',
                    adapter: {
                        type: 'adapter',
                    },
                    hooks: {},
                    issues: {
                        title: 'issue title',
                        description: 'issue description',
                        state: 'open',
                        state_reason: 'not_planned',
                        labels: ['bug']
                    }
                },
                workingDirectory: 'workingDirectory',
                selectedRepos: [{ name: 'selectedRepos' }],
                repos: [{ name: 'selectedRepos' }],
                upstreamOwner: 'upstreamOwner',
            },
            adapter: mockAdapter,
            logger: mockLogger,
        };
    });

    it('should list issues when the command is invoked',
        async () => {

            (getIssueTrackerFile as jest.Mock).mockResolvedValueOnce(
                [{
                    issueNumber: '7',
                    title: 'this is my first updated issue',
                    owner: 'newOwner',
                    repo: 'newRepo'
                }]
            );

            await listIssues(mockContext);

            expect(process.stdout.write).toEqual(expect.any(Function))
        });

});
