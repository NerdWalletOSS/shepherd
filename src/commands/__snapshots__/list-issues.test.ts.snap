// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`list-issue command should list issues when the command is invoked 1`] = `
[MockFunction] {
  "calls": [
    [
      "╔══════════════╤═════════════╤════════════╤════════╤═══════════╗
║ issue Number │ issue Title │ Owner      │ status │ Repo Name ║
╟──────────────┼─────────────┼────────────┼────────┼───────────╢
║ 3            │ new title   │ Nerdwallet │ closed │ shepherd  ║
╚══════════════╧═════════════╧════════════╧════════╧═══════════╝
",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": true,
    },
  ],
}
`;
